docs/basics/101-113-symlinks.rst

```diff
@@ -0,0 +1,222 @@
+   Understanding more about the object tree can help to understand such
+   problems, and knowing bits of the Git-annex basics can make you more
+   confident in working with your datasets.
+
+   So how do these paths and names come into existence?
+
+   When a file is annexed, Git-annex generates a *key* from the file content
+   that it uses (in part) as a name for the file and as a part of the path
+   in the object tree.
+   The key is associated with the content of the file (the *value*).
+   Therefore, using this key, file content can be (somewhat) identified
+   (or rather: Based on the keys, it can be identified whether two files
+   have identical contents, or whether file content changed).
+
+   The key is generated using *hashes*. A hash is a function that turns an
+   input (e.g. a PDF file) into a string of characters with a fixed length.
+   In principle, therefore, the hash function simply transforms a content of
+   any size into a string with fixed length.
+
+   The important aspect of a hash function is that it
+   will generate the same hash for the same file content, but once file content
+   changes, the generated hash will also look differently. If two files are
+   turned into identical character strings, the content in these files is thus
+   identical.
+
+   This key (or :term:`checksum`) is the last part of the name of the file the
+   symlink links to (in which the actual data content
+   is stored). The extension (e.g. ``.pdf``) is appended because some
+   operating systems need this information. The key is also one of the subdirectory
+   names in the path. This two-level structure is implemented because it helps to prevent
+   accidental deletions and changes.
+
+   .. runrecord:: _examples/DL-101-113-104
+      :language: console
+      :workdir: dl-101/DataLad-101/books
+
+      # take a look at the last part of the target path:
+      $ ls -lah TLCL.pdf
+
+   .. runrecord:: _examples/DL-101-113-105
+      :language: console
+      :workdir: dl-101/DataLad-101/books
+
+      # compare it to the checksum (here of type md5sum) of the PDF file and the subdirectory name
+      $ md5sum TLCL.pdf
+
+   There are different hash functions available. Depending on which is used,
+   the resulting :term:`checksum` has a certain length and structure.
+   By default, DataLad uses ``MD5e`` checksums, but should you want to, you can
+   change this default to `one of many other types <https://git-annex.branchable.com/backends/>`_.
+   The first part of the file name actually states which hash function is used.
+
+   By now we know where almost all parts of the file name derived from - the remaining
+   unidentified bit in the file name is the
+   one after the checksum identifier. This part is the size of the content in bytes. An annexed
+   file in the object tree thus has a file name following this structure:
+
+   ``checksum-identifier - size -- checksum . extension``
```

Comment:
Side note: URL-type keys can come without the size info, see here https://github.com/datalad-datasets/machinelearning-books/blob/master/H.DaumeIII-A_Course_in_Machine_Learning.pdf

As far as I understand, this can only happen, if data lives on a server that does not report the size of a file in the HTPP response to a request from git-annex. It should be an exotic exception, but it can happen: https://github.com/datalad/datalad/issues/3519